File: src/pt/joaomneto/titancompanion/adventure/impl/TCOCAdventure.java
package pt.joaomneto.titancompanion.adventure.impl;

import java.io.BufferedWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import pt.joaomneto.titancompanion.R;
import pt.joaomneto.titancompanion.adventure.Adventure;

import android.os.Bundle;
import android.view.Menu;

public class TCOCAdventure extends Adventure {

    Map<String, Integer> spells = new HashMap<>();
    List<String> spellList = new ArrayList<>();
    Integer spellValue;

    protected static final int FRAGMENT_SPELLS = 2;
    protected static final int FRAGMENT_EQUIPMENT = 3;
    protected static final int FRAGMENT_NOTES = 4;

    public TCOCAdventure() {
        super();
        fragmentConfiguration.clear();
        fragmentConfiguration
                .put(FRAGMENT_VITAL_STATS,
                        new AdventureFragmentRunner(R.string.vitalStats,
                                "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureVitalStatsFragment"));
        fragmentConfiguration
                .put(FRAGMENT_COMBAT,
                        new AdventureFragmentRunner(R.string.fights,
                                "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureCombatFragment"));
        fragmentConfiguration
                .put(FRAGMENT_SPELLS,
                        new AdventureFragmentRunner(
                                R.string.spells,
                                "pt.joaomneto.titancompanion.adventure.impl.fragments.tcoc.TCOCAdventureSpellsFragment"));
        fragmentConfiguration
                .put(FRAGMENT_EQUIPMENT,
                        new AdventureFragmentRunner(R.string.goldEquipment,
                                "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureEquipmentFragment"));
        fragmentConfiguration
                .put(FRAGMENT_NOTES,
                        new AdventureFragmentRunner(R.string.notes,
                                "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureNotesFragment"));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        try {
            super.onCreate(savedInstanceState);

        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.adventure, menu);
        return true;
    }

    @Override
    public void storeAdventureSpecificValuesInFile(BufferedWriter bw)
            throws IOException {

        String spellsS = "";

        if (!spells.isEmpty()) {
            for (String spell : getSpellList()) {
                spellsS += spell + "§"+spells.get(spell)+"#";
            }
            spellsS = spellsS.substring(0, spellsS.length() - 1);
        }

        bw.write("spells=" + spellsS + "\n");
        bw.write("spellValue=" + spellValue + "\n");
        bw.write("gold=" + getGold() + "\n");
    }


    @Override
    protected void loadAdventureSpecificValuesFromFile() {
        setGold(Integer.valueOf(getSavedGame().getProperty("gold")));
        spellValue = Integer.valueOf(getSavedGame().getProperty("spellValue"));
        String spellsS = new String(getSavedGame().getProperty("spells")
                .getBytes(java.nio.charset.Charset.forName("UTF-8")));

        if (spellsS != null) {
            spells = new HashMap<>();
            List<String> list = Arrays.asList(spellsS.split("#"));
            for (String string : list) {
                if (!string.isEmpty()) {
                    String[] split = string.split("§");
                    spells.put(split[0], Integer.parseInt(split[1]));
                }
            }

        }

    }

    public synchronized List<String> getSpellList() {
        for (String spell : spells.keySet()) {
            if (!spellList.contains(spell)) {
                spellList.add(spell);
            }
        }

        Collections.sort(spellList);
        return spellList;
    }


    public void removeSpell(int position) {
        String spell = getSpellList().get(position);
        int value = spells.get(spell) - 1;
        if (value == 0) {
            spells.remove(spell);
            getSpellList().remove(spell);
        } else {
            spells.put(spell, value);
        }
    }

    public Map<String, Integer> getSpells() {
        return spells;
    }

    public int getSpellListSize() {
        int size = 0;
        for (String spell : getSpellList()) {
            Integer value = spells.get(spell);
            if (value != null) {
                size += value;
            } else {
                size += 1;
            }
        }
        return size;
    }

}

File: src/main/java/pt/joaomneto/titancompanion/adventure/impl/TCOCAdventure.kt
package pt.joaomneto.titancompanion.adventure.impl

import android.os.Bundle
import android.view.Menu
import pt.joaomneto.titancompanion.R
import pt.joaomneto.titancompanion.adventure.Adventure
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureCombatFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureEquipmentFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureNotesFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureVitalStatsFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.tcoc.TCOCAdventureSpellsFragment
import pt.joaomneto.titancompanion.util.AdventureFragmentRunner
import java.io.BufferedWriter
import java.io.IOException
import java.util.ArrayList
import java.util.Arrays
import java.util.Collections
import java.util.HashMap

class TCOCAdventure : Adventure(
    arrayOf(
        AdventureFragmentRunner(
            R.string.vitalStats,
            AdventureVitalStatsFragment::class
        ),
        AdventureFragmentRunner(
            R.string.fights,
            AdventureCombatFragment::class
        ),
        AdventureFragmentRunner(
            R.string.spells,
            TCOCAdventureSpellsFragment::class
        ),
        AdventureFragmentRunner(
            R.string.goldEquipment,
            AdventureEquipmentFragment::class
        ),
        AdventureFragmentRunner(
            R.string.notes,
            AdventureNotesFragment::class
        )
    )
) {

    internal var spells: MutableMap<String, Int> = HashMap()
    internal var spellList: MutableList<String> = ArrayList()
    internal var spellValue: Int? = null

    val spellListSize: Int
        get() {
            var size = 0
            for (spell in getSpellList()) {
                val value = spells[spell]
                if (value != null) {
                    size += value
                } else {
                    size += 1
                }
            }
            return size
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        try {
            super.onCreate(savedInstanceState)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.adventure, menu)
        return true
    }

    @Throws(IOException::class)
    override fun storeAdventureSpecificValuesInFile(bw: BufferedWriter) {

        var spellsS = ""

        if (!spells.isEmpty()) {
            for (spell in getSpellList()) {
                spellsS += spell + "§" + spells[spell] + "#"
            }
            spellsS = spellsS.substring(0, spellsS.length - 1)
        }

        bw.write("spells=" + spellsS + "\n")
        bw.write("spellValue=" + spellValue + "\n")
        bw.write("gold=" + gold + "\n")
    }

    override fun loadAdventureSpecificValuesFromFile() {
        gold = Integer.valueOf(savedGame.getProperty("gold"))
        spellValue = Integer.valueOf(savedGame.getProperty("spellValue"))
        val spellsS = String(
            savedGame.getProperty("spells")
                .toByteArray(java.nio.charset.Charset.forName("UTF-8"))
        )

        spells = HashMap()
        val list = Arrays.asList(*spellsS.split("#".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
        for (string in list) {
            if (!string.isEmpty()) {
                val split = string.split("§".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                spells.put(split[0], Integer.parseInt(split[1]))
            }
        }
    }

    @Synchronized
    fun getSpellList(): MutableList<String> {
        for (spell in spells.keys) {
            if (!spellList.contains(spell)) {
                spellList.add(spell)
            }
        }

        Collections.sort(spellList)
        return spellList
    }

    fun removeSpell(position: Int) {
        val spell = getSpellList()[position]
        val value = spells[spell]!! - 1
        if (value == 0) {
            spells.remove(spell)
            getSpellList().remove(spell)
        } else {
            spells.put(spell, value)
        }
    }

    fun getSpells(): Map<String, Int> {
        return spells
    }

    companion object {

        protected val FRAGMENT_SPELLS = 2
        protected val FRAGMENT_EQUIPMENT = 3
        protected val FRAGMENT_NOTES = 4
    }
}
